keps/sig-storage/20190621-namespace-transfer-api.md

```diff
@@ -0,0 +1,230 @@
+
+While the concept of transferring an object across namespaces is not necessarily something that fits for "all" resource types, there are a number of storage objects today that this would be useful; volumes, snapshots, clones etc.  In addition, the number of data related objects in kubernetes is expected to grow, and include things like external populators, backups and other content sources.  It's important to plan ahead and provide a shared API that can be used the same way regardless of object type if and when it makes sense and they have a valid use case for transfer functionality.
+
+## Motivation
+
+* Provide a standard API for the transfer of kubernetes objects from one namespace to another
+
+### Goals
+
+* Standardized API and process for performing a namespace transfer regardless of object
+* Enable implementation for performing a namespace transfer for core objects
+* Enable implementation for performing a namespace transfer for external (CRD) objects
+* Define a mechanism to indicate whether an object implements the namespace transfer API
+
+### Non-Goals
+
+* Actual implementation of transferring an object between namespaces
+
+## Proposal
+
+Introduce an external `NamespaceTransferRequest` object:
+
+```yaml
+apiversion: v1alpha1
+kind: NamespaceTransferRequest
+metadata:
+    name: pvc-transfer-request
+    namespace:  destination-namespace
+spec:
+    source:
+        name: source-namespace
+        name: source-pvc
+        kind: PersistentVolumeClaim
+
+```
+
+Introduce an external `NamespaceTransferApproval` object:
+
+```yaml
+apiversion: v1alpha1
+kind: NamespaceTransferApproval
+metadata:
+    name: pvc-transfer-approval
+    namespace: source-namespace
+spec: 
+    source:
+        name: source-pvc
+        kind: PersistentVolumeClaim
+    targetNamespace: destination-namespace
+
+```
+
+Introduce a parameter to objects (including the existing `kubernetes.pkg.apis.core.types.go.ObjectMeta` structure 
+to indicate if an object is transferable:
+
+```go
+type ObjectMeta struct {
+      ...
+      Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+	  // Transferable indicates whether the object implements the namespace transfer API.  By default
+	  // the controller responsible for creation of the resource will set this appropriately based on
+	  // whether or not it implements the namespace transfer API.  A user may also explicitly set this
+	  // to `false` if they wish to prevent any transfer operations against the particular resource being
+	  // created regardless of the controllers capabilities.
+	  // +optional
+	  Transferable bool `json:"export,omitempty" protobuf:"varint,1,opt,name=transferable"`
```

Comment:
The bar for this is super high.  It would be much better if you demonstrated the use case and solution for this by creating PVC specific objects with a focused controller.  Moving objects across namespaces is not a significantly requested action to justify a new object meta feature, and it’s always better to start from a working implementation for a specific use case.

It’s also likely storage has domain specific fields that users would want, so starting with a generic action is too risky.

I’m not in favor of a generic object proposal until a specific use case is widely deployed.